Support multiple IPv4 netmasks per interface

jukkar opened the pull request:
The netmask should be tied to the IPv4 address instead of being global for the network interface.

If there is only one IPv4 address specified to the network interface, nothing changes from user point of view. But if there are more than one IPv4 address / network interface, the netmask must be specified to each address separately.

This means that the `net_if_ipv4_get_netmask()` and `net_if_ipv4_set_netmask()` functions should not be used as they will only work reliably if there is only one IPv4 address in the network interface.

Instead new `net_if_ipv4_get_netmask_by_addr()` and `net_if_ipv4_set_netmask_by_addr()` should be used instead as they make sure that the netmask is tied to correct IPv4 address in the network interface.

No need to try to rush and try to get this to 3.6, existing functionality works for most of the use cases.

jukkar committed:
net: ipv4: Make netmask IPv4 address specific

The netmask should be tied to the IPv4 address instead of being
global for the network interface.

If there is only one IPv4 address specified to the network interface,
nothing changes from user point of view. But if there are more than
one IPv4 address / network interface, the netmask must be specified
to each address separately.

This means that net_if_ipv4_get_netmask() and net_if_ipv4_set_netmask()
functions should not be used as they only work reliably if there is
only one IPv4 address in the network interface.

The new net_if_ipv4_get_netmask_by_addr() and
net_if_ipv4_set_netmask_by_addr() functions should be used as they make
sure that the netmask is tied to correct IPv4 address in the network
interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: ipv4: Fix tests to support new netmask handling

Make sure that the tests work properly when each IPv4 address
in the network interface gets its own netmask.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: ipv4: Deprecate the old netmask set/get routines

This commit deprecates these legacy netmask get/set routines

net_if_ipv4_set_netmask()
net_if_ipv4_set_netmask_by_index()
net_if_ipv4_get_netmask()

as they do not work well if there are multiple IPv4 address
assigned to the network interface.

User should use these functions instead

net_if_ipv4_set_netmask_by_addr()
net_if_ipv4_set_netmask_by_addr_by_index()
net_if_ipv4_get_netmask_by_addr()

as they make sure the netmask it bound to correct IPv4 address.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos reviewed (commented):
Looks good, thanks, howerver we should get rid of those now deprecated functions usage within the codebase.
One of the users is recently added DHCPv4 server:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/dhcpv4/dhcpv4_server.c#L1234
but in this particular case it should be straightforward fix, just take the netmask for the selected `server_addr` .

jukkar committed:
drivers: Change drivers to support new IPv4 netmask setting API

Make sure that the drivers use the new IPv4 netmask setting API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: Change samples to support new IPv4 netmask setting API

Make sure that the samples use the new IPv4 netmask setting API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4_server.c:
```diff
@@ -1231,7 +1231,7 @@ int net_dhcpv4_server_start(struct net_if *iface, struct in_addr *base_addr)
 	}
 
-	netmask = net_if_ipv4_get_netmask(iface);
+	netmask = net_if_ipv4_get_netmask_by_addr(iface, base_addr);
```
I think `server_addr` should be used here instead. `base_addr` won't be configured as IPv4 address on the interface.

jukkar replied:
Indeed, will fix.

jukkar committed:
net: if: Convert array loops to use ARRAY_FOR_EACH macro

Convert various array loops in the net_if.c to use the
ARRAY_FOR_EACH() macro. This makes the code more robust
as we do not need to keep track of the separate define
that tells the array size.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: mgmt: hawkbit: Simplify DHCPv4 address detection

No need to loop through network interface addresses, just
listen DHCPv4 bound event and use that as a condition to continue.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Convert various address loops to use ARRAY_FOR_EACH macro

Convert various networking subsystem files to use ARRAY_FOR_EACH
macro to make the looping more robust.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

aescolar commented:
Setting milestone as 3.7 as we are past the RC1 for 3.6. Please correct the milestone if you disagree

pdgendt commented on include/zephyr/net/net_if.h:
```diff
@@ -368,19 +368,28 @@ struct net_if_dhcpv6 {
+ *
+ * Stores the unicast IPv4 address and related netmask.
+ */
+struct net_if_addr_ipv4 {
+	/** IPv4 address */
+	struct net_if_addr ipv4;
+	/** Netmask */
+	struct in_addr netmask;
+};
```
If a netmask is always a bit mask of a certain length `111...00` (which I think it is), you could have the netmask length set in the `_unused` part of the `struct net_if_addr`.

pdgendt replied:
This is in line with the `<ip>/<netmask len>` notation, for example `192.168.0.1/24`.

jukkar replied:
We could store the length too, but I went this route so calculating the netmask beforehand and then using it directly. With length, one would need to calculate the netmask for the matching operation. We loose three bytes, netmask is 4 bytes long, vs. one byte for the length when storing the data.

pdgendt replied:
It still fits in the unused 5 bits, but more importantly, the interface struct doesn't break with out-of-tree usage.

jukkar replied:
Sure, everything is possible of course, but I suggest that we go with this proposal atm. If you feel strong about this, it is possible to change how this works in subsequent PRs.

pdgendt replied:
Not that strongly opiniated, but I'm always cautious when modifying (public) header files. If we accept the proposal no "revert" PR should be added.

pdgendt replied:
CC @rlubos

rlubos replied:
I don't have a strong opinion on this, there are certainly various ways/places where netmask could be stored alongside the address, current approach definitely works and seems rather straightforward in use.

But perhaps we should indeed mention this change in the migration guide?

jukkar replied:
> But perhaps we should indeed mention this change in the migration guide?

Yes, that was my plan too. Will do it now when we have a migration doc for 3.7.
